### ISSUE TITLE
Guide(integrate): Login to Proxmox

### ISSUE THREAD

**Issue by mffap** [zitadel/zitadel#8055]
### Preflight Checklist

- [X] I could not find a solution in the existing issues, docs, nor discussions
- [ ] I have joined the [ZITADEL chat](https://zitadel.com/chat)

### Describe the docs your are missing or that are wrong

As proxmox user I want to configure ZITADEL IdP, so that I can authenticate to proxmox

### Additional Context

Proxmox seems a popular tool, here some arguments why a guide could be helpful for the community: 
- High popularity OS among self-hosting: https://selfh.st/survey/2023-results/
- Authentik has useful docs, according to users: https://docs.goauthentik.io/integrations/services/proxmox-ve/
- We see some search demand on our docs (no results)
- Proxmox seems to be also a topic in our community chat from time to time

It might come down to some standard configuration, but the community might benefit from a more detailed guide.

**Comment by ojsef39** [comment]
Yeah, please add that guide :)

I've tried multiple times now to set up Proxmox with Zitadel and failed, so an official documentation on how it's done from Zitadels side would be very cool :)

**Comment by maxcerny** [comment]
Is there any progress on this? Because right now it looks like it's not possible to integrate proxmox and zitadel.

**Comment by alexmoras** [comment]
At the moment I don't believe its possible due to an issue with how Proxmox parses `client_id`'s. They strip the `@project_name` part hence why it shows the "invalid audience" error. Until Proxmox fix this, or Zitadel adds an option to set alphanumeric-only client ID's, there is little purpose for adding a guide.

**Comment by maxcerny** [comment]
Yeah, that's what saw as well. I wonder if I can force proxmox to not strip it...

**Comment by alexmoras** [comment]
There is a bug report for Proxmox here - https://bugzilla.proxmox.com/show_bug.cgi?id=5076.

I've just updated it to now.

You can update the parsing regex to support the "@" symbol but any updates to that part of the Proxmox core will break authentication again.

**Comment by maxcerny** [comment]
I am probably blind but, could you point me to the regex? I can't find it...

**Comment by alexmoras** [comment]
I can't seem to find it either now - the Proxmox codebase is a maze. I've just noticed the pull request #8178 which provides functionality to remove the "@" from the client ID. It's been merged in to main. Just trying to see if its made its way in to any of the docker images.

**Comment by alexmoras** [comment]
It seems the docker images [v2.56.0-rc.4](https://github.com/orgs/zitadel/packages/container/zitadel/239653823?tag=v2.56.0-rc.4) onwards all have the new code which removes the suffix from Client IDs. You need to create a new app for it to apply, all old ones will keep the suffix.

**Comment by maxcerny** [comment]
> I can't seem to find it either now - the Proxmox codebase is a maze.

It really is...

I'll probably wait for the release, don't really have the need for it to work right now. Great to see this has moved somewhere :)

**Comment by hifabienne** [comment]
We just merged a pr from @btajuddin which removes the @ and project id when generating the client id. https://github.com/zitadel/zitadel/pull/8178
It will be included in the next release.
After that it should be possible from what I understand.

**Comment by Aurel004** [comment]
Did anyone managed to get it working and could provide a simple guide ?

I am having a error 500

**Comment by maxcerny** [comment]
I didn't, swapped to authentik. Had other issues with zitadel as well, so I just didn't bother.

**Comment by reynold-lariza** [comment]
@maxcerny so basically Zitadel is not usable?

**Comment by maxcerny** [comment]
> @maxcerny so basically Zitadel is not usable?

Not sure, but there were some talks about changing how the client IDs work. If they went through, then it should be possible. But I haven't followed it, so idk how that ended up.

You can try it out, just be sure to exclude client IDs suffix in zitadel if possible.

**Comment by reynold-lariza** [comment]
@maxcerny Just tried the whole night, trial and error - but still failed. I'm running the latest Proxmox 8.2.4 with the latest docker container of Zitadel. And there's not much tutorial or guidance.

Swapped to Authentik, and boy it worked OOTB!

**Comment by maxcerny** [comment]
@reynold-lariza yeah same, also had way more features, such as an LDAP outpost

**Comment by katosdev** [comment]
Curious, what were you using for your issuer URL? 
I have this working on Authentik and am looking to replicate over on Zitadel as I'd ideally like to move my services to this.

**Comment by HearthCore** [comment]
Hey there, also looking at Zitadel for my MSP Project.
ProxMox not working with OICD of Zitadel is a blocker for us.

**Comment by isodude** [comment]
Issue tracking multiple audiences: https://github.com/zitadel/zitadel/issues/9200

**Comment by hifabienne** [comment]
Input from Proxmox Developer:
- A fix is implemented, but still needs some review and be merged, after that will go into the next release.


Steps/ Docs to follow to set everything up (Input from Proxmox Developer):


# SETTING UP ZITADEL WITH PROXMOX VE

## CREATING A ZITADEL INSTANCE

To create and use a Zitadel instance, there are two approaches one can take: i) installing Zitadel on your system without Docker and ii) setting up an instance of Zitadel with Docker. To do the former, please read the official Zitadel documentation [here](https://zitadel.com/docs/self-hosting/deploy/linux). To take the latter route, please read the official documentation [here](https://zitadel.com/docs/self-hosting/deploy/compose).

If you are using a custom domain, please take note of the following important points and options in your `docker-compose.yaml`:

- To configure the `ExternalDomain` option, add the environment variable `ZITADEL_EXTERNALDOMAIN` inside the `environment` section of the `zitadel` service and set it to your server's IP address or domain if you own one.
- To configure the `ExternalPort` option, add the environment variable `ZITADEL_EXTERNALPORT` inside the environment section of the `zitadel` service and set it to an unused port of your machine.

After you have created your instance, you can visit your Zitadel instance via this URL: `your-IP:Port/ui/console`. The placeholder `your-IP` stands for the IP address of the server on which you are running Zitadel and the one you used to fill the `ZITADEL_EXTERNALDOMAIN` field in your Docker Compose file. The placeholder `Port` stands for the port you used to fill the `ZITADEL_EXTERNALPORT` field in your Docker Compose file.

## ADDING ZITADEL AUTH TO YOUR PROXMOX VE INSTANCE

### Configuring Zitadel

Read the official Zitadel documentation located [here](https://zitadel.com/docs/guides/integrate/services/cloudflare-oidc) and follow all steps until the "Review your configuration" step. Inside Zitadel, set the redirect URL to the domain or IP address of your Proxmox instance.

### Adding Zitadel to Proxmox VE

Please follow these steps to add Zitadel as an Open ID authentication provider to your Proxmox VE instance:

- Go to the "Datacenter" button on the left side of the screen.
- Go to the "Realms" option.
- Click on the button that says "Add".
- Choose "OpenID Connect".
- In the "Issuer URL" field, put the IP address or domain of your Zitadel instance.
- In the "Client ID" field, put the number sequence Zitadel presented you with as your client ID after you created your first Zitadel app.
- In the "Client Secret" field, put the random character sequence Zitadel presented you with as your client secret after you created your first Zitadel app.
- Check the "Autocreate Users" box.
- Inside Zitadel, go to the page of the project you created and copy the long number under the label "Resource ID".
- Inside Proxmox, paste this number into the "Audiences" field.
- Return to Zitadel and go to the page of the app you created and copy the long number under the label "Resource ID".
- Inside Proxmox, paste this number into the "Audiences" field.
- Make sure the numbers of the project ID and the app ID are separated by spaces inside the Audience field.

**Comment by ojsef39** [comment]
> ## ADDING ZITADEL AUTH TO YOUR PROXMOX VE INSTANCE


Thanks for updating this, but i can tell you without trying it again that this wont work (if there werent changes in zitadel)
I did it exactly like this when i initially commented here, and that didnt work.

If there were any chances to Zitadel which make this possible now, please let me know and ill try again.

**Comment by ojsef39** [comment]
> Input from Proxmox Developer:
> 
> * A fix is implemented, but still needs some review and be merged, after that will go into the next release.
> 

Thank you for you update on this, any chance you could keep us up to date when this is gonna be merged by proxmox?

**Comment by ojsef39** [comment]
> Unfortunately, I don't know exactly when that will be. Here is the link to the bug at Proxmox. When the state is changed to applied, it should be included. https://bugzilla.proxmox.com/show_bug.cgi?id=5076

Thanks! Ill keep an eye out

**Comment by ibnunowshad** [comment]
Audience field in Proxmox? I do not see any such field in Proxmox

**Comment by hifabienne** [comment]
From looking at the bug report at proxmox, it shows that the patch is available, version 8.
As far as I can see the current proxmox version is 8.4, which would mean, it should now be possible to integrate Zitadel with Proxmox.

Did someone already test this and can confirm?

**Comment by ibnunowshad** [comment]
> From looking at the bug report at proxmox, it shows that the patch is available, version 8. As far as I can see the current proxmox version is 8.4, which would mean, it should now be possible to integrate Zitadel with Proxmox.
> 
> Did someone already test this and can confirm?

I was in `8.3.5` when I tested it and I do not see any `audience` field. I am updating `8.4` now. I will test with my capacity and update it here.

**Comment by TheTaran** [comment]
I also have my Proxmox cluster running on version 8.4.1, but OpenID with ZITADEL still isn't working. I get the error:
`msg=Failed to verify ID token: Invalid audiences: '###########' is not a trusted audience.`
Is there any update on when this will be fixed?

**Comment by hifabienne** [comment]
Did you also follow the instructions posted here? https://github.com/zitadel/zitadel/issues/8055#issuecomment-2626798965

As this fix was done on the side of proxmox, I can't tell you when this will work

**Comment by TheTaran** [comment]
Yes, I configured this. But I don't have any field named "Audiences".

![Image](https://github.com/user-attachments/assets/a3c4d2e9-2f73-4d43-b725-5aba9d72fcdf)

**Comment by isodude** [comment]
From what I can tell it has not landed here https://git.proxmox.com/?p=proxmox.git;a=blob;f=proxmox-openid/src/lib.rs;h=4f3c7c18791ee4b460570922df2cfa841a570d7b;hb=HEAD yet.

**Comment by ghost** [comment]
I'm actually the developer at Proxmox who fixed this bug and wrote the guide above. I'm close to finishing this series of patches but I still need to send a v3 and I am awaiting review. If you would like to test the v2 or v3 of this patch series, it would help to send a mail to our PVE-devel mailing list with a trailer of "Tested-by: <you@email.com>".

The link to the v2 is here: https://lore.proxmox.com/all/999303783.1030.1748874882155@webmail.proxmox.com/
The link to the bug in our Bugzilla is here: https://bugzilla.proxmox.com/show_bug.cgi?id=5076

(Edit: v3 sent)

**Comment by TheTaran** [comment]
Updated today `proxmox to latest release 9.0.5, The problem is steel present. Any news?

**Comment by SoFMeRight** [comment]
I emailed the patch list hoping maybe I could be instructed how to apply and test the  patch... Don't wanna nag, but it would be really cool. I have the lack of sanity to push this into home-prod if the capability was provided. I'm half tempted to try and see if claude code can figure out how to apply the patch from the provided source cause I'm clueless about PVE's source and how to patch/build isolated bits lol. Hopefully we have an update on this soon. No rush I definitely check on this issue every few weeks for maybe longer than a year but I understand beggers can't be choosers I dont even have a license lol 🙃 Thankful its even being developed at all!!!!

**Comment by ilbarone87** [comment]
Any update here? Anyone knows if the patch has been implemented into proxmox 9.1

**Comment by Dr-Shadow** [comment]
@ilbarone87 it seems the proxmox staff which was working on our patch left the company and is now read as "Proxmox retired staff"

EDIT : the proxmox staff is now aware about this and they should have a look on it

**Comment by nakamorichi** [comment]
any update?

**Comment by Dr-Shadow** [comment]
Hello @nakamorichi 

You should check for updates there : https://bugzilla.proxmox.com/show_bug.cgi?id=5076

**Comment by Dr-Shadow** [comment]
It seems there is a new assignee on the Proxmox bugzilla ticket

Let's keep fingers crossed !

**Comment by vsviridov** [comment]
Looks like things are happening code-wise:

https://lore.proxmox.com/pve-devel/20260423133548.349086-1-s.sterz@proxmox.com/T/#t

**Comment by ilbarone87** [comment]
Any news if this will land in 9.2?

**Comment by Dr-Shadow** [comment]
Changes have been made in libpve-access-control on 9.1.1 version and other packages.

I can confirm the audiences field is available in the realm OpenID Connect configuration form after update.

I had success using it.

To Zitadel Team : this issue could be closed

**Comment by Vazgen005** [comment]
How to change Username Claim?
<img width="667" height="505" alt="Image" src="https://github.com/user-attachments/assets/044aaaff-873a-4649-ad43-5d5095e04ca9" />

**Comment by Vazgen005** [comment]
Seems to it supposed to be editable 
https://pve.proxmox.com/pve-docs/chapter-pveum.html#_username_mapping

**Comment by ilbarone87** [comment]
I’m trying to map the group from zitadel so the new user is added automatically to the group in zitadel. The action pass the group claims in my zitadel. Has anyone tried this too?

**Comment by SoFMeRight** [comment]
I previously struggled with understanding why the instructions detailed in the guide in this thread Login to Proxmox were not suffice to get this working in my environment. 

I kept getting the issue of proxmox rejecting audience IDs that I couldnt find in the app I setup for PVE OIDC:
- Audiences should be typed as a comma separated list without spaces
- In my case I use Zitadel and it required for the Project ID and *all* Client ID's within it to be listed 
> (not just the client ID for a application you added to the project for Proxmox SSO specifically, ALL of the apps in that Project, I repeat that again; my project had portainer and other stuff in it; I immediately zoned it to a project with just PBS and PVE when I figured this out for a total of 4 audiences IDs required)

The other issue I see another user @ibnunowshad has above is mapping groups:
- I use a "Actions" within the application in Zitadel to flatten the roles/groups to "roles" and return just their values, by default Zitadel nests them into a more complicated structure, I do this for most apps it helps make it easier. I think they put a guide on how to do this where I copy pasted this block from in the Zitadel docs (I am not 100% sure thats where I got it, but Im sure it would be easy to find in a search engine.)
```
function flatRoles(ctx, api) {
  if (ctx.v1.user.grants == undefined || ctx.v1.user.grants.count == 0) {
    return;
  }

  let grants = [];
  ctx.v1.user.grants.grants.forEach(claim => {
    claim.roles.forEach(role => {
        grants.push(role)  
    })
  })

  api.v1.claims.setClaim('roles', grants)
}
```

This absolutely works. I am thankful to the developers and everyone else who bumped this issue, reminded them after the past developer left not to forget to get this feature shipped 🎉 
I think this issue can be closed? Just a thought. Thank you so much again everyone!
